### DISCUSSION TITLE
Tower PDK

### DISCUSSION THREAD

**Discussion by luomeng007** [gdsfactory/gdsfactory#3422] [Q&A]
I use Tower PDK to plot my Layout. However, the package they provide is GDSFactory 7. Now we use GDSFactory 8. A lot of functions can not be used. I want to ask how to overcome this? Like the code below:

`import tj
import gdsfactory as gf

xs = gf.cross_section.strip_heater_metal_undercut(width=0.5, heater_width=2, trench_width=4, trench_gap=4)
p = gf.path.arc(radius=10, angle=45)
c = p.extrude(xs)
c.plot()`

And when I run the code, I get error like this.

> ValueError                                Traceback (most recent call last)
> Cell In[3], [line 6](vscode-notebook-cell:?execution_count=3&line=6)
>       [4](vscode-notebook-cell:?execution_count=3&line=4) xs = gf.cross_section.strip_heater_metal_undercut(width=0.5, heater_width=2, trench_width=4, trench_gap=4)
>       [5](vscode-notebook-cell:?execution_count=3&line=5) p = gf.path.arc(radius=10, angle=45)
> ----> [6](vscode-notebook-cell:?execution_count=3&line=6) c = p.extrude(xs)
>       [7](vscode-notebook-cell:?execution_count=3&line=7) c.plot()
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\gdsfactory\path.py:468, in Path.extrude(self, cross_section, layer, width, widths, simplify, shear_angle_start, shear_angle_end)
>     [432](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:432) def extrude(
>     [433](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:433)     self,
>     [434](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:434)     cross_section: CrossSectionSpec | None = None,
>    (...)
>     [440](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:440)     shear_angle_end: float | None = None,
>     [441](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:441) ) -> Component:
>     [442](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:442)     """Returns Component by extruding a Path with a CrossSection.
>     [443](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:443) 
>     [444](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:444)     A path can be extruded using any CrossSection returning a Component
>    (...)
>     [466](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:466)         c.plot()
>     [467](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:467)     """
> --> [468](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:468)     return extrude(
>     [469](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:469)         p=self,
>     [470](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:470)         cross_section=cross_section,
>     [471](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:471)         layer=layer,
>     [472](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:472)         width=width,
>     [473](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:473)         widths=widths,
>     [474](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:474)         simplify=simplify,
>     [475](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:475)         shear_angle_start=shear_angle_start,
>     [476](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:476)         shear_angle_end=shear_angle_end,
>     [477](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:477)     )
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\gdsfactory\cell.py:199, in cell_without_validator.<locals>._cell(*args, **kwargs)
>     [194](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:194) if not callable(func):
>     [195](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:195)     raise ValueError(
>     [196](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:196)         f"{func!r} is not callable! @cell decorator is only for functions"
>     [197](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:197)     )
> --> [199](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:199) component = func(*args, **kwargs)
>     [201](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:201) # if the component is already in the cache, but under a different alias,
>     [202](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:202) # make sure we use a copy, so we don't run into mutability errors
>     [203](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/cell.py:203) if id(component) in [id(v) for v in CACHE.values()]:
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\pydantic\decorator.py:40, in pydantic.decorator.validate_arguments.validate.wrapper_function()
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\pydantic\decorator.py:134, in pydantic.decorator.ValidatedFunction.call()
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\pydantic\decorator.py:206, in pydantic.decorator.ValidatedFunction.execute()
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\gdsfactory\path.py:906, in extrude(p, cross_section, layer, width, widths, simplify, shear_angle_start, shear_angle_end)
>     [904](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:904) width = section.width
>     [905](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:905) offset = section.offset
> --> [906](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:906) layer = get_layer(section.layer)
>     [907](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:907) port_names = section.port_names
>     [908](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/path.py:908) port_types = section.port_types
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\gdsfactory\pdk.py:747, in get_layer(layer)
>     [746](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:746) def get_layer(layer: LayerSpec) -> Layer:
> --> [747](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:747)     return get_active_pdk().get_layer(layer)
> 
> File c:\Users\luome\Desktop\GDSFactory\HGXS_Layout\Lib\site-packages\gdsfactory\pdk.py:591, in Pdk.get_layer(self, layer)
>     [589](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:589) elif isinstance(layer, str):
>     [590](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:590)     if layer not in self.layers:
> --> [591](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:591)         raise ValueError(f"{layer!r} not in {self.layers.keys()}")
>     [592](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:592)     return self.layers[layer]
>     [593](file:///C:/Users/luome/Desktop/GDSFactory/HGXS_Layout/Lib/site-packages/gdsfactory/pdk.py:593) elif layer is np.nan:
> 
> ValueError: 'DEEPTRENCH' not in dict_keys(['cni', 'contact', 'contact_slot', 'cpi', 'dsvholep', 'geano', 'geca', 'geop', 'geano2', 'ipmark', 'ltrench', 'metal1_block', 'metal1boundary', 'metal_mirror', 'metal1', 'metal1_fill', 'metal1_ind', 'metal1_pin', 'metal1_resistor', 'metal1_slot', 'metal1_tline', 'metal2_block', 'metal2_boundary', 'metal2', 'metal2_fill', 'metal2_ind', 'metal2_pin', 'metal2_resistor', 'metal2_slot', 'metal2_tline', 'metal3_block', 'metal3_boundary', 'metal3', 'metal3_fill', 'metal3_ind', 'metal3_pin', 'metal3_resistor', 'metal3_slot', 'metal3_tline', 'modano2', 'modano', 'modca2', 'modca', 'm_CUPad', 'm_L_cell', 'm_PIN', 'm_SRmarkR', 'm_ablb', 'm_analog', 'm_artifact', 'm_bumpPad', 'm_dataExtent', 'm_diodev', 'm_esd', 'm_grtcpl', 'm_pd', 'm_ph_arc', 'm_ph_bend', 'm_ph_cross', 'm_ph_dc', 'm_ph_mmi', 'm_ph_mzi', 'm_ph_str', 'm_ph_yjunct', 'm_restmrk', 'm_wgcnct', 'nwg_block', 'nwg', 'nwg_fill', 'nwg_pin', 'ng2', 'outline', 'prBoundary', 'resdev', 'revgeo', 'revnwg', 'revsblk', 'revswg', 'rwg', 'rwg_pin', 'sblk', 'scrbo', 'silox', 'sus', 'swg_block', 'swg', 'swg_fill', 'swg_pin', 'gwg', 'text', 'tnr', 'topmm', 'vg1', 'via1', 'via1slot', 'via2', 'via2slot', 'vgroove_open', 'vgroove_hardmask', 'ubm', 'ubm2', 'label', 'fiber_ec', 'dicing', 'fiber_gc', 'padding', 'PADDING', 'IR_temp', 'LABEL_INSTANCE', 'DEVREC', 'PORT', 'PORTE', 'TE', 'TM', 'TEXT', 'LABEL', 'WG', 'SHOW_PORTS', 'HEATER', 'swg_passivation', 'wafer'])

**Comment by joamatab** [comment]
For foundry support, please reach out to us at contact@gdsfactory.com.

**Reply by luomeng007** [reply]
I have sent e-mail to you
